File: android/app/src/main/java/com/voxtv/firetv/OpenSubtitles.kt
package com.voxtv.firetv

import java.io.ByteArrayInputStream
import java.nio.ByteBuffer
import java.nio.charset.Charset
import java.nio.charset.CodingErrorAction
import java.util.zip.GZIPInputStream
import okhttp3.Request
import org.json.JSONArray
import org.json.JSONObject

/**
 * Port of `src/lib/iptv/opensubtitles.ts`. Same endpoints, same filtering and
 * the same SubRip → WebVTT conversion, so the subtitle sheet behaves exactly as
 * it does in the browser build.
 */
object OpenSubtitles {
  private const val SEARCH_HOST = "https://rest.opensubtitles.org/search"
  private const val OS_USER_AGENT = "TemporaryUserAgent"
  private const val MAX_RESULTS = 12
  private const val MAX_SUBTITLE_BYTES = 2 * 1024 * 1024

  private val NOISE = Regex(
    "\\b(1080p|720p|480p|2160p|4k|uhd|hdr|web-?dl|webrip|bluray|brrip|hdrip|dvdrip|x264|x265|h264|h265|hevc|aac|ac3|dts|hdtv|multi|dual|sub|vo?stfr)\\b",
    RegexOption.IGNORE_CASE,
  )

  fun normalizeQuery(raw: String): String =
    raw
      .replace(Regex("\\.(mkv|mp4|avi|ts|m3u8)$", RegexOption.IGNORE_CASE), "")
      .replace(NOISE, " ")
      .replace(Regex("[\\[\\]()_.\\-–—:|]+"), " ")
      .replace(Regex("\\s+"), " ")
      .trim()
      .take(120)

  fun isLang(value: String): Boolean = value == "eng" || value == "ara"

  fun langLabel(lang: String): String = if (lang == "ara") "العربية" else "English"

  fun buildSearchUrl(query: String, langs: List<String>, season: Int?, episode: Int?): String {
    val segments = ArrayList<String>()
    if (episode != null && episode > 0) segments.add("episode-$episode")
    segments.add("query-${Http.encodeComponent(query)}")
    if (season != null && season > 0) segments.add("season-$season")
    val wanted = if (langs.isEmpty()) listOf("eng") else langs.distinct()
    segments.add("sublanguageid-${wanted.joinToString(",")}")
    return "$SEARCH_HOST/${segments.joinToString("/")}"
  }

  private fun request(url: String, accept: String): okhttp3.Response {
    val req = Request.Builder()
      .url(url)
      .header("X-User-Agent", OS_USER_AGENT)
      .header("User-Agent", OS_USER_AGENT)
      .header("Accept", accept)
      .build()
    return IptvProxy.client.newCall(req).execute()
  }

  fun search(query: String, langs: List<String>, season: Int?, episode: Int?): ServerResponse {
    val normalized = normalizeQuery(query)
    if (normalized.isEmpty()) return ServerResponse.text(200, """{"results":[]}""", "application/json")
    val url = buildSearchUrl(normalized, langs, season, episode)
    val body: String
    try {
      request(url, "application/json").use { res ->
        if (!res.isSuccessful) {
          val message = if (res.code == 429) {
            "OpenSubtitles is rate limiting requests. Try again in a minute."
          } else {
            "OpenSubtitles search failed (${res.code})."
          }
          return ServerResponse.text(502, """{"error":${Json.quote(message)}}""", "application/json")
        }
        body = res.body?.string() ?: "[]"
      }
    } catch (err: Exception) {
      return ServerResponse.text(
        502,
        """{"error":${Json.quote("Could not reach OpenSubtitles. Check your connection and try again.")}}""",
        "application/json",
      )
    }
    val results = parseResults(body, langs)
    return ServerResponse.text(200, """{"results":$results}""", "application/json")
  }

  private fun parseResults(payload: String, langs: List<String>): String {
    val rows = try {
      JSONArray(payload)
    } catch (_: Exception) {
      return "[]"
    }
    val wanted = langs.toSet()
    val buckets = LinkedHashMap<String, MutableList<JSONObject>>()
    for (i in 0 until rows.length()) {
      val row = rows.optJSONObject(i) ?: continue
      val lang = row.optString("SubLanguageID").lowercase()
      val id = row.optString("IDSubtitleFile")
      if (id.isEmpty() || !isLang(lang) || !wanted.contains(lang)) continue
      val format = row.optString("SubFormat", "srt").lowercase()
      if (format != "srt" && format != "vtt") continue
      val release = row.optString("MovieReleaseName")
      val name = listOf(row.optString("SubFileName"), release, row.optString("MovieName"))
        .firstOrNull { it.isNotEmpty() } ?: "Subtitle $id"
      val hit = JSONObject()
        .put("id", id)
        .put("name", name)
        .put("lang", lang)
        .put("langLabel", langLabel(lang))
        .put("format", format)
        .put("downloads", row.optString("SubDownloadsCnt", "0").toIntOrNull() ?: 0)
        .put("rating", row.optString("SubRating", "0").toDoubleOrNull() ?: 0.0)
        .put("downloadUrl", row.optString("SubDownloadLink"))
      if (release.isNotEmpty()) hit.put("release", release)
      buckets.getOrPut(lang) { ArrayList() }.add(hit)
    }
    val sorted = buckets.values.map { bucket ->
      bucket.sortedWith(
        compareByDescending<JSONObject> { it.optInt("downloads") }.thenByDescending { it.optDouble("rating") },
      )
    }
    val ordered = JSONArray()
    var index = 0
    while (ordered.length() < MAX_RESULTS) {
      var added = false
      for (bucket in sorted) {
        if (index >= bucket.size) continue
        ordered.put(bucket[index])
        added = true
        if (ordered.length() >= MAX_RESULTS) break
      }
      if (!added) break
      index++
    }
    return ordered.toString()
  }

  private fun isOpenSubtitlesUrl(url: String): Boolean = try {
    val host = java.net.URI(url).host?.lowercase() ?: ""
    host == "opensubtitles.org" || host.endsWith(".opensubtitles.org")
  } catch (_: Exception) {
    false
  }

  fun download(id: String, lang: String, url: String?): ServerResponse {
    val candidates = ArrayList<String>()
    if (!url.isNullOrBlank() && isOpenSubtitlesUrl(url)) candidates.add(url)
    candidates.add("https://dl.opensubtitles.org/en/download/subencoding-utf8/file/${Http.encodeComponent(id)}")

    var lastError = "Could not download that subtitle."
    for (candidate in candidates) {
      try {
        request(candidate, "*/*").use { res ->
          if (!res.isSuccessful) {
            lastError = "OpenSubtitles download failed (${res.code})."
            return@use
          }
          val raw = res.body?.bytes() ?: ByteArray(0)
          if (raw.size > MAX_SUBTITLE_BYTES) {
            lastError = "That subtitle file is too large."
            return@use
          }
          val bytes = if (looksGzipped(raw)) gunzip(raw) else raw
          val vtt = srtToVtt(decodeSubtitle(bytes, lang))
          if (!vtt.contains("-->")) {
            lastError = "That subtitle file could not be read."
            return@use
          }
          return ServerResponse.text(200, vtt, "text/vtt")
        }
      } catch (err: Exception) {
        lastError = "Could not reach OpenSubtitles. Check your connection and try again."
      }
    }
    return ServerResponse.text(502, """{"error":${Json.quote(lastError)}}""", "application/json")
  }

  fun looksGzipped(bytes: ByteArray): Boolean =
    bytes.size > 2 && bytes[0] == 0x1f.toByte() && bytes[1] == 0x8b.toByte()

  private fun gunzip(bytes: ByteArray): ByteArray =
    GZIPInputStream(ByteArrayInputStream(bytes)).use { it.readBytes() }

  /** UTF-8 when it decodes cleanly, otherwise the usual Latin/Arabic codepages. */
  fun decodeSubtitle(bytes: ByteArray, lang: String): String {
    val strict = Charsets.UTF_8.newDecoder()
      .onMalformedInput(CodingErrorAction.REPORT)
      .onUnmappableCharacter(CodingErrorAction.REPORT)
    try {
      return strict.decode(ByteBuffer.wrap(bytes)).toString()
    } catch (_: Exception) {
      // not valid UTF-8 — fall through to the legacy codepages
    }
    val fallback = if (lang == "ara") "windows-1256" else "windows-1252"
    return try {
      String(bytes, Charset.forName(fallback))
    } catch (_: Exception) {
      String(bytes, Charsets.UTF_8)
    }
  }

  /** Drops SSA/ASS override blocks and <font> tags WebVTT cannot render. */
  fun cleanCueText(text: String): String =
    text
      .replace(Regex("\\{\\\\[^}]*\\}"), "")
      .replace(Regex("</?font[^>]*>", RegexOption.IGNORE_CASE), "")
      .lines().joinToString("\n") { it.trimEnd() }
      .trim()

  fun srtToVtt(input: String): String {
    val text = input.removePrefix("﻿").replace(Regex("\r\n?"), "\n").trim()
    if (text.isEmpty()) return "WEBVTT\n\n"
    val body = if (text.startsWith("WEBVTT")) text.substring(6).trimStart() else text
    val cues = ArrayList<String>()
    for (block in body.split(Regex("\n{2,}"))) {
      val lines = block.split("\n").filter { it.isNotBlank() }
      if (lines.isEmpty()) continue
      var index = 0
      if (lines[0].trim().matches(Regex("^\\d+$")) && lines.size > 1 && lines[1].contains("-->")) index = 1
      val timing = lines.getOrNull(index) ?: continue
      if (!timing.contains("-->")) continue
      val parts = timing.split("-->")
      if (parts.size < 2) continue
      val start = parts[0].trim().replace(',', '.')
      val end = parts[1].trim().split(Regex("\\s+")).firstOrNull()?.replace(',', '.') ?: continue
      val payload = cleanCueText(lines.drop(index + 1).joinToString("\n"))
      if (payload.isEmpty()) continue
      cues.add("$start --> $end\n$payload")
    }
    return "WEBVTT\n\n${cues.joinToString("\n\n")}\n"
  }
}

File: src/lib/iptv/opensubtitles.ts
/**
 * OpenSubtitles.org lookup used when a stream carries no usable subtitle
 * track. Runs on the server (the API blocks browser origins and ships
 * gzipped SRT files), and only ever deals with English and Arabic.
 */
import { iptvLog, iptvWarn } from "./log.ts";
import {
  isSubtitleLang,
  langLabel,
  type SubtitleHit,
  type SubtitleLang,
} from "./subtitle-types.ts";

export type { SubtitleHit, SubtitleLang };

const SEARCH_HOST = "https://rest.opensubtitles.org/search";
/** The legacy REST endpoint rejects requests without this header. */
const OS_USER_AGENT = "TemporaryUserAgent";
const MAX_RESULTS = 12;
const MAX_SUBTITLE_BYTES = 2 * 1024 * 1024;

/** Strip release noise so "Movie.Name.2019.1080p.WEB" still matches. */
export function normalizeQuery(raw: string): string {
  return raw
    .replace(/\.(mkv|mp4|avi|ts|m3u8)$/i, "")
    .replace(
      /\b(1080p|720p|480p|2160p|4k|uhd|hdr|web-?dl|webrip|bluray|brrip|hdrip|dvdrip|x264|x265|h264|h265|hevc|aac|ac3|dts|hdtv|multi|dual|sub|vo?stfr)\b/gi,
      " ",
    )
    .replace(/[[\]()_.\-–—:|]+/g, " ")
    .replace(/\s+/g, " ")
    .trim()
    .slice(0, 120);
}

/**
 * Legacy REST paths are made of alphabetically ordered `key-value` segments.
 */
export function buildSearchPath(opts: {
  query: string;
  langs: SubtitleLang[];
  season?: number;
  episode?: number;
}): string {
  const segments: string[] = [];
  if (opts.episode && opts.episode > 0) segments.push(`episode-${opts.episode}`);
  segments.push(`query-${encodeURIComponent(opts.query)}`);
  if (opts.season && opts.season > 0) segments.push(`season-${opts.season}`);
  const langs = opts.langs.length ? opts.langs : ["eng" as SubtitleLang];
  segments.push(`sublanguageid-${[...new Set(langs)].join(",")}`);
  return `${SEARCH_HOST}/${segments.join("/")}`;
}

interface RawHit {
  IDSubtitleFile?: string;
  SubFileName?: string;
  SubLanguageID?: string;
  LanguageName?: string;
  SubFormat?: string;
  SubDownloadsCnt?: string | number;
  SubRating?: string | number;
  SubDownloadLink?: string;
  MovieReleaseName?: string;
  MovieName?: string;
}

export function parseSearchResults(payload: unknown, langs: SubtitleLang[]): SubtitleHit[] {
  if (!Array.isArray(payload)) return [];
  const wanted = new Set(langs);
  const hits: SubtitleHit[] = [];
  for (const row of payload as RawHit[]) {
    const lang = String(row?.SubLanguageID ?? "").toLowerCase();
    const id = String(row?.IDSubtitleFile ?? "");
    if (!id || !isSubtitleLang(lang) || !wanted.has(lang)) continue;
    const format = String(row?.SubFormat ?? "srt").toLowerCase();
    if (format !== "srt" && format !== "vtt") continue;
    hits.push({
      id,
      name: String(row?.SubFileName || row?.MovieReleaseName || row?.MovieName || `Subtitle ${id}`),
      lang,
      langLabel: langLabel(lang),
      format,
      downloads: Number(row?.SubDownloadsCnt ?? 0) || 0,
      rating: Number(row?.SubRating ?? 0) || 0,
      downloadUrl: String(row?.SubDownloadLink ?? ""),
      release: row?.MovieReleaseName ? String(row.MovieReleaseName) : undefined,
    });
  }
  const byLang = new Map<SubtitleLang, SubtitleHit[]>();
  for (const hit of hits) {
    const bucket = byLang.get(hit.lang) ?? [];
    bucket.push(hit);
    byLang.set(hit.lang, bucket);
  }
  // Interleave languages so a mixed search never buries one of them.
  const ordered: SubtitleHit[] = [];
  const buckets = [...byLang.values()].map((bucket) =>
    bucket.sort((a, b) => b.downloads - a.downloads || b.rating - a.rating),
  );
  for (let i = 0; ordered.length < MAX_RESULTS; i++) {
    let added = false;
    for (const bucket of buckets) {
      const hit = bucket[i];
      if (!hit) continue;
      ordered.push(hit);
      added = true;
      if (ordered.length >= MAX_RESULTS) break;
    }
    if (!added) break;
  }
  return ordered;
}

async function osFetch(url: string, accept: string, timeoutMs: number): Promise<Response> {
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), timeoutMs);
  try {
    return await fetchOrThrow(url, accept, ctrl.signal);
  } finally {
    clearTimeout(timer);
  }
}

async function fetchOrThrow(url: string, accept: string, signal: AbortSignal): Promise<Response> {
  try {
    return await fetch(url, {
      headers: {
        "X-User-Agent": OS_USER_AGENT,
        "User-Agent": OS_USER_AGENT,
        Accept: accept,
        "Accept-Encoding": "gzip, identity",
      },
      redirect: "follow",
      signal,
    });
  } catch (err) {
    iptvWarn("subs", "network", err instanceof Error ? err.message : err);
    throw new Error("Could not reach OpenSubtitles. Check your connection and try again.");
  }
}

export async function searchSubtitles(opts: {
  query: string;
  langs: SubtitleLang[];
  season?: number;
  episode?: number;
}): Promise<SubtitleHit[]> {
  const query = normalizeQuery(opts.query);
  if (!query) return [];
  const url = buildSearchPath({ ...opts, query });
  iptvLog("subs", "search", query, opts.langs.join(","));
  const res = await osFetch(url, "application/json", 15000);
  if (!res.ok) {
    iptvWarn("subs", "search failed", res.status);
    throw new Error(
      res.status === 429
        ? "OpenSubtitles is rate limiting requests. Try again in a minute."
        : `OpenSubtitles search failed (${res.status}).`,
    );
  }
  const payload = (await res.json().catch(() => null)) as unknown;
  const hits = parseSearchResults(payload, opts.langs);
  iptvLog("subs", "search results", hits.length);
  return hits;
}

function isOpenSubtitlesUrl(url: string): boolean {
  try {
    const host = new URL(url).hostname.toLowerCase();
    return host === "opensubtitles.org" || host.endsWith(".opensubtitles.org");
  } catch {
    return false;
  }
}

async function gunzip(bytes: Uint8Array): Promise<Uint8Array> {
  const stream = new Blob([bytes as BlobPart])
    .stream()
    .pipeThrough(new DecompressionStream("gzip"));
  const buf = await new Response(stream).arrayBuffer();
  return new Uint8Array(buf);
}

export function looksGzipped(bytes: Uint8Array): boolean {
  return bytes.length > 2 && bytes[0] === 0x1f && bytes[1] === 0x8b;
}

/** UTF-8 when it decodes cleanly, otherwise the usual Latin/Arabic codepages. */
export function decodeSubtitleBytes(bytes: Uint8Array, lang: SubtitleLang): string {
  try {
    return new TextDecoder("utf-8", { fatal: true }).decode(bytes);
  } catch {
    const fallback = lang === "ara" ? "windows-1256" : "windows-1252";
    try {
      return new TextDecoder(fallback).decode(bytes);
    } catch {
      return new TextDecoder("utf-8").decode(bytes);
    }
  }
}

function vttTimestamp(raw: string): string {
  const value = raw.trim().replace(",", ".");
  // WebVTT needs at least mm:ss.mmm; SRT always ships hh:mm:ss,mmm.
  return value;
}

/**
 * SRT files in the wild carry SSA/ASS override blocks and <font> tags that
 * WebVTT does not understand and would render as literal text.
 */
export function cleanCueText(text: string): string {
  return text
    .replace(/\{\\[^}]*\}/g, "")
    .replace(/<\/?font[^>]*>/gi, "")
    .replace(/[ \t]+$/gm, "")
    .trim();
}

/** Convert SubRip (or pass through WebVTT) into a cue list the browser accepts. */
export function srtToVtt(input: string): string {
  const text = input
    .replace(/^\uFEFF/, "")
    .replace(/\r\n?/g, "\n")
    .trim();
  if (!text) return "WEBVTT\n\n";
  const body = text.startsWith("WEBVTT") ? text.slice(6).trimStart() : text;
  const blocks = body.split(/\n{2,}/);
  const cues: string[] = [];
  for (const block of blocks) {
    const lines = block.split("\n").filter((line) => line.trim() !== "");
    if (!lines.length) continue;
    let index = 0;
    if (/^\d+$/.test(lines[0]!.trim()) && lines[1]?.includes("-->")) index = 1;
    const timing = lines[index];
    if (!timing || !timing.includes("-->")) continue;
    const [start, rest] = timing.split("-->");
    if (!start || !rest) continue;
    const end = rest.trim().split(/\s+/)[0] ?? "";
    const payload = cleanCueText(lines.slice(index + 1).join("\n"));
    if (!payload) continue;
    cues.push(`${vttTimestamp(start)} --> ${vttTimestamp(end)}\n${payload}`);
  }
  return `WEBVTT\n\n${cues.join("\n\n")}\n`;
}

export async function downloadSubtitle(opts: {
  id: string;
  lang: SubtitleLang;
  url?: string;
}): Promise<string> {
  const candidates = [
    opts.url && isOpenSubtitlesUrl(opts.url) ? opts.url : "",
    `https://dl.opensubtitles.org/en/download/subencoding-utf8/file/${encodeURIComponent(opts.id)}`,
  ].filter(Boolean);

  let lastError = "Could not download that subtitle.";
  for (const url of candidates) {
    try {
      const res = await osFetch(url, "*/*", 20000);
      if (!res.ok) {
        lastError = `OpenSubtitles download failed (${res.status}).`;
        continue;
      }
      const buf = new Uint8Array(await res.arrayBuffer());
      if (buf.byteLength > MAX_SUBTITLE_BYTES) {
        lastError = "That subtitle file is too large.";
        continue;
      }
      const raw = looksGzipped(buf) ? await gunzip(buf) : buf;
      const text = decodeSubtitleBytes(raw, opts.lang);
      const vtt = srtToVtt(text);
      if (!/-->/.test(vtt)) {
        lastError = "That subtitle file could not be read.";
        continue;
      }
      iptvLog("subs", "download ok", opts.id, vtt.length);
      return vtt;
    } catch (err) {
      lastError = err instanceof Error ? err.message : lastError;
      iptvWarn("subs", "download failed", lastError);
    }
  }
  throw new Error(lastError);
}
